corellium/adapter/src/main/kotlin/flank/corellium/adapter/InvokeAndroidDevices.kt

```diff
@@ -0,0 +1,137 @@
+package flank.corellium.adapter
+
+import flank.corellium.api.AndroidInstance
+import flank.corellium.api.AndroidInstance.Config
+import flank.corellium.client.core.createNewInstance
+import flank.corellium.client.core.getAllProjects
+import flank.corellium.client.core.getProjectInstancesList
+import flank.corellium.client.core.startInstance
+import flank.corellium.client.core.waitUntilInstanceIsReady
+import flank.corellium.client.data.Instance
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.joinAll
+import kotlinx.coroutines.launch
+
+private const val FLANK_INSTANCE_NAME_PREFIX = "flank-android-"
+private const val FLAVOUR = "ranchu"
+private const val OS = "11.0.0"
+private const val SCREEN = "720x1280:280"
+
+class InvokeAndroidDevices(
+    private val projectName: String,
+) : AndroidInstance.Invoke {
+    override suspend fun Config.invoke(): List<String> = coroutineScope {
+        val projectId = getProjectId(projectName)
+        val instances = getCreatedInstances(projectId, amount)
+        startNotRunningInstances(instances)
+
+        val ids = instances.map(Instance::id) + let {
+            // When existing instances size match required amount
+            // there is not needs for creating more instances.
+            if (instances.size == amount) emptyList()
+            // Otherwise is required to create some additional instances
+            else createInstances(
+                projectId = projectId,
+                indexes = calculateAdditionalIndexes(
+                    current = instances,
+                    requiredAmount = amount
+                )
+            )
+        }
+
+        waitForInstances(ids)
+        ids
+    }
+}
+
+// Important!!!
+// Try to keep the private methods not dependent on each other.
+// Otherwise the implementation will go complicated.
+
+/**
+ * @return The project id for given project name.
+ */
+private suspend fun getProjectId(name: String) =
+    corellium.getAllProjects().first { it.name == name }.id
+
+/**
+ * Get all instances that was already created for flank.
+ * @return [List] of [Instance] where [List.size] <= [amount]
+ */
+private suspend fun getCreatedInstances(
+    projectId: String,
+    amount: Int
+): List<Instance> = corellium
+    .also { println("Getting instances already created by flank.") }
+    .getProjectInstancesList(projectId)
+    .filter { it.name.startsWith(FLANK_INSTANCE_NAME_PREFIX) }
+    .filter { it.state !in Instance.State.unavailable }
+    .take(amount)
+    .apply { println("Obtained $size already created devices") }
+
+/**
+ * Start all given instances with status different than "on".
+ */
+private suspend fun startNotRunningInstances(
+    instances: List<Instance>
+): Unit = instances
+    .filter { it.state != "on" }
```

Comment:
It might be worth making "on" a constant as its used in multiple places.
Up to you :)

Reply:
I see its been defined already, can it be used here?